spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala

```diff
@@ -330,25 +335,49 @@ trait DeltaColumnMappingBase extends DeltaLogging {
   def assignColumnIdAndPhysicalName(
       newMetadata: Metadata,
       oldMetadata: Metadata,
-      isChangingModeOnExistingTable: Boolean): Metadata = {
+      isChangingModeOnExistingTable: Boolean,
+      isOverwritingSchema: Boolean): Metadata = {
     val rawSchema = newMetadata.schema
     var maxId = DeltaConfigs.COLUMN_MAPPING_MAX_ID.fromMetaData(newMetadata) max
                 findMaxColumnId(rawSchema)
     val newSchema =
       SchemaMergingUtils.transformColumns(rawSchema)((path, field, _) => {
-        val builder = new MetadataBuilder()
-          .withMetadata(field.metadata)
+        val builder = new MetadataBuilder().withMetadata(field.metadata)
+
+        lazy val fullName = path :+ field.name
+        lazy val existingFieldOpt =
+          SchemaUtils.findNestedFieldIgnoreCase(
+            oldMetadata.schema, fullName, includeCollections = true)
+        lazy val canReuseColumnMappingMetadataDuringOverwrite = {
+          val canReuse =
+            isOverwritingSchema &&
+              SparkSession.getActiveSession.exists(
+                _.conf.get(DeltaSQLConf.REUSE_COLUMN_MAPPING_METADATA_DURING_OVERWRITE)) &&
+              existingFieldOpt.exists { existingField =>
+                // Ensure data type & nullability are compatible
+                DataType.equalsIgnoreCompatibleNullability(existingField.dataType, field.dataType)
```

Comment:
It's not clear from the function call that this is not a symmetric comparison. And even if it were clear, it's then unclear which is the "from" and which is the "to" type.
```suggestion
                DataType.equalsIgnoreCompatibleNullability(from = existingField.dataType, to = field.dataType)
```